docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+
+I rewrote the application without MobX and using only using React's context API and useEffect to respond to state changes. The application state remained in a hierarchy, but without any observer/observable pattern.  The state was passed down the component hierarchy via context.  Container components detected state changes via additional useEffect calls. In order to coordinate larger reset of the application state (such as clearing most data), an application pub/sub mechanism was used to communicate between components to orchestrate the sequence of clearing state. This was because the useEffect order was difficult to guarantee.
+
+Finally, I rewrote the application to use Recoil. I flattened the application state into atoms. Because relationships between the data create a logical hierarchy, I wrote utility functions to make cross referencing between data islands easy. I then wrote selectors to shape the data for my React components. The container components leveraged useRecoilValue and useRecoilState to respond to state changes. Because the same application state change be updated from different React components, I implemented a dispatcher pattern based on useRecoilCallback.  I had several missteps as I learned how Recoil works, and discovered how my application was affected by flattening the state.
+
+# Recoil API: Pros, Cons, Unknowns, and Hazards
+
+## Overall
+- PRO: Built for hooks from the start and follows React concepts
+- PRO: Atom and selectors allow for composing applications from small pieces without introducing unnecessary dependencies.
+- PRO: Significantly reduces the complexity and code volume in components.
+- PRO: Looking toward the future of asynchronous, multi-threaded Javascript.
+- PRO: Recoil API surface is tiny
+- PRO: Lightning fast (at least for my application size)
+- CON: Still experimental, few releases, limited contributes - not yet mature or fully tested.
+
+## Atoms
+- PRO: Like their name, atoms let you model your data as small and independent plain types.
+- PRO: Serializing atoms to storage is straightforward compared to trying to rehydrate an observed state tree.
+- PRO: The default state of an atom with the reset() hook is convenient for clearing application state.
+- CON: Atoms are often arrays of items which means needing to look up items. This requires more undefined checks for cases where items are not found.
```

Comment:
Can this be addressed via a selector?

Reply:
Yes you can use a selector to handle if for example a list is empty, however if you wanted to look up an element in that list by a unique parameter the constraint still exists

Reply:
Sometimes you can return an empty value when something isn't found, you can enforce something is always there when setting the atom or throwing if there isn't . Anytime you have an array and are using findIndex or find, there is the chance something isn't there.

Reply:
Yes. Unfortunately, Typescript does not yet support something like `@throws` so I'm not sure if we can ever not leak the undefined case out of a selector.